### DISCUSSION TITLE
Passing additional attributes to Javascript Policies using AuthzClient

### DISCUSSION THREAD

**Discussion by dsmith-to** [keycloak/keycloak#31785] [Q&A]
We have Java services that use AuthzClient to retrieve permission for a user.  Until now we have been using primarily for RBAC policies. 

We have a new ABAC requirements for which we have enable Javascript policies.  These policies require addition context from the service to evaluate the permissions.  Specifically,  we want to pass a company_id and grant access based on users inclusion in company-specific roles.  

I was led to believe (by a chatbot) that we could provided additional context attributes to the JS-policy using AuthorizationRequest.setClaims  but this doesn't seem to be the case.  Is there any way to do this?

**Comment by dsmith-to** [comment]
I changed the code to MAC sign a claims token.   I'm signing with the services client secret and is sent from the resource server, but the user's token is issued for a public client.
Result is 403   Public clients are not allowed to send claims

Any way I can do this?

**Comment by sachingaikwad123** [comment]
Hi @dsmith-to,

Did you solve this problem? I am also trying to ABAC using my own custom javascript policy and I would like to pass extra 'attributes' to keycloak which will be used in evaluation. How to do this?

These attributes are dynamic - meaning. Depending on instance of 'resource', it can be different. Resource Server will know this attribute from 'REST API request' sent by client. Resource Server will extract these "attributes" and pass it to Keycloak (PDP).

**Comment by imrabti** [comment]
You could achieve that using Keycloak Policy enforcer, and enhance the REST endpoint with “ Claim Information Point”. I have done that and the endpoint will be called automatically to retrieve additional.

A sample how to add Claim Information Point: 

```
{
        "path": "/movies/{id}/rent",
        "methods": [
          {
            "method": "POST",
            "scopes": ["movies-app:scope:rent"]
          },
          {
            "method": "DELETE",
            "scopes": ["movies-app:scope:giveBack"]
          }
        ],
        "claim-information-point": {
          "http": {
            "url": "http://localhost:7010/movies/rented",
            "method": "GET",
            "headers": {
              "Authorization": "Bearer {keycloak.access_token}",
              "Content-Type": "application/json"
            }
          }
        }
      }
```

And the in a JavaScript Policy you can access this data and put it you policy business logic, below a sample code of JS Policy that uses the data supplied by the Claim Information Point:

```
var context = $evaluation.getContext();
var identity = context.getIdentity();
var attributes = identity.getAttributes();

if (attributes.containsValue('blackListed', 'true')) {
  $evaluation.deny();
} else {
  if (attributes.containsValue('vip', 'true')) {
    $evaluation.grant();
  } else {
    // Notice here the attribute the was loaded from the Claim Information Point by called the rest service
    var rentedMoviesCount = context.attributes.getValue('rentedMoviesCount').asLong(0);
    if (rentedMoviesCount > 0) {
        $evaluation.deny();
    } else {
      $evaluation.grant();
    }
  }
}
```

I hope this could help you find a solution for your requirement.
